nfs/src/rpc_task.cpp

```diff
@@ -2772,27 +2805,107 @@ static void readdir_callback(
+                                  task->rpc_api->readdir_task.get_target_offset());
+                    }
                 } else {
                     assert((int64_t) dircache_handle->get_eof_cookie() != -1);
                 }
             }
         }
 
-        task->send_readdir_response(readdirentries);
+        // Only send to fuse if we have seen new entries or EOF.
+        if(should_send_response || eof) {
+            task->send_readdir_response(readdirentries);
+        }
+
     } else if (NFS_STATUS(res) == NFS3ERR_JUKEBOX) {
         task->get_client()->jukebox_retry(task);
+        return;
+    } else if (NFS_STATUS(res) == NFS3ERR_BAD_COOKIE) {
+        AZLogWarn("[{}] got NFS3ERR_BAD_COOKIE for offset: {}, cleaning "
+                  "dircache and starting reenumeration", 
+                  dir_ino, task->rpc_api->readdir_task.get_offset());
+        dir_inode->invalidate_cache();
+
+        /*
+         * We have received a bad cookie error, we have to restart enumeration 
+         * until either the server returns a valid response or we reach eof. 
+         * If we keep getting bad cookie we will keep on reenumerating forever.
+         */
+        last_valid_offset = 0;
+        
+        /*
+         * We need to maintain the monotonocity of the target_offset
+         * because it represents the offsets already sent to fuse as part of 
+         * this enumeration. This protects us from sending duplicate entries
+         * to fuse if we receive bad_cookie before we reach the target during
+         * reenumeration. 
+         */
+        task->rpc_api->readdir_task.set_target_offset(
+                                    std::max(task->rpc_api->readdir_task.get_offset() + 1, 
+                                    task->rpc_api->readdir_task.get_target_offset()));
     } else {
         task->reply_error(status);
+        return;
+    }
+
+ 
+    /*
+     * We have not seen a new entry and the call has not failed hence, this is a
+     * reenumeration call and we have not reached the target_offset yet. We have to
+     * start another readdir call for the next batch. 
+     */
+    if (!should_send_response) {
+        struct rpc_task *child_tsk =
+            task->get_client()->get_rpc_task_helper()->alloc_rpc_task(FUSE_READDIR);
+        
+        child_tsk->init_readdir(
+            task->rpc_api->req,
+            task->rpc_api->readdir_task.get_ino(),
+            task->rpc_api->readdir_task.get_size(),
+            last_valid_offset,
+            task->rpc_api->readdir_task.get_target_offset(),
+            task->rpc_api->readdir_task.get_fuse_file());
+
+        /*
+         * Set the parent task of the child to the parent of the
+         * current RPC task. This is required if the current task itself
+         * is one of the child tasks running part of the request.
+         */
+        child_tsk->rpc_api->parent_task = task->rpc_api->parent_task;
+        
```

Comment:
unlike read, here we don't use parent_task for tracking ongoing readdirs, so parent_task will be null and we don't need to copy it